### PR TITLE
fix:[CDS-48469]: Added the retry mechanism

### PR DESCRIPTION
**Pull request by shivamnegi94**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41804)

### CODE REVIEW

**Review by vivekkarnatiharness — APPROVED** [review]
> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41804)* status: 3 file(s) to review. 1 of 2 approvals obtained.

**Review by abhinav-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41804)* status: 3 file(s) to review. 1 change requests. 1 of 2 approvals obtained. 

___
*[`960-api-services/src/main/java/io/harness/artifacts/gcr/service/GcrApiServiceImpl.java` line 113 at r2](https://reviewable.io/reviews/harness/harness-core/41804#-NJl0EIF9my24hV4IyM5:-NJl0EIF9my24hV4IyM6:birhrug) ([raw file](https://github.com/harness/harness-core/blob/84deb70813ae239c60c02fe88b010dc2069a241e/960-api-services/src/main/java/io/harness/artifacts/gcr/service/GcrApiServiceImpl.java#L113)):*
> ```Java
>       throw handleIOException(gcpConfig, e);
>     } catch (InvalidRequestException e) {
>       throw new HintException(e.getMessage());
> ```

ExceptionUtils.getmessage

**Review by shivamnegi94 — COMMENTED** [review]
> Reviewed 1 of 3 files at r1, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41804)* status: 3 file(s) to review. 1 change requests. 0 of 2 approvals obtained.

**Review by shivamnegi94 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41804)* status: 3 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`960-api-services/src/main/java/io/harness/artifacts/gcr/service/GcrApiServiceImpl.java` line 113 at r2](https://reviewable.io/reviews/harness/harness-core/41804#-NJl0EIF9my24hV4IyM5:-NJmZG7ACwJabFe0n8lF:bs7wumb) ([raw file](https://github.com/harness/harness-core/blob/84deb70813ae239c60c02fe88b010dc2069a241e/960-api-services/src/main/java/io/harness/artifacts/gcr/service/GcrApiServiceImpl.java#L113)):*
<details><summary><i>Previously, abhinav-harness (Abhinav Singh) wrote…</i></summary><blockquote>

ExceptionUtils.getmessage
</blockquote></details>

Done.

**Review by abhinav-harness — DISMISSED** [review]
> Reviewed 1 of 3 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41804)* status: 1 file(s) to review. 1 of 2 approvals obtained.

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41804)* status: 2 file(s) to review. 1 of 2 approvals obtained.

**Review by abhinav-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41804#-:-NJrr_HD72qH3GhkeumG:bnfp4nl)* status: 2 file(s) to review. 1 of 2 approvals obtained.

File: 960-api-services/src/main/java/io/harness/artifacts/gcr/service/GcrApiServiceImpl.java [hunk 1]
```diff
@@ -238,6 +241,19 @@ public Response<DockerImageManifestResponse> fetchImage(
...
   }
 
+  @VisibleForTesting
+  public Response listImageTag(GcrRestClient gcrRestClient, String basicAuthHeader, String imageName) throws Exception {
+    return Retry.decorateCallable(retry, () -> gcrRestClient.listImageTags(basicAuthHeader, imageName).execute())
+        .call();
+  }
+
+  @VisibleForTesting
+  public Response<GcrImageTagResponse> getBuilds(GcrRestClient gcrRestClient, String basicAuthHeader, String imageName)
```
Comment (sribalij) [inline 960-api-services/src/main/java/io/harness/artifacts/gcr/service/GcrApiServiceImpl.java:251]: This method `getBuilds(GcrRestClient gcrRestClient, String basicAuthHeader, String imageName)` and
`listImageTag(GcrRestClient gcrRestClient, String basicAuthHeader, String imageName) throws Exception {`
are doing the same thing right except for return type?

File: 960-api-services/src/main/java/io/harness/artifacts/gcr/service/GcrApiServiceImpl.java [hunk 2]
```diff
@@ -238,6 +241,19 @@ public Response<DockerImageManifestResponse> fetchImage(
...
+      throws Exception {
+    return Retry.decorateCallable(retry, () -> gcrRestClient.listImageTags(basicAuthHeader, imageName).execute())
```
Comment (sribalij) [inline 960-api-services/src/main/java/io/harness/artifacts/gcr/service/GcrApiServiceImpl.java:253]: This is also going to retry only once? There should be a way to specify the retry config?

Comment (shivamnegi94) [reply 960-api-services/src/main/java/io/harness/artifacts/gcr/service/GcrApiServiceImpl.java:253]: We are retrying it 5 times, we are initialising that in constructor.
` public GcrApiServiceImpl() {
    final RetryConfig config =
        RetryConfig.custom().maxAttempts(5).intervalFunction(IntervalFunction.ofExponentialBackoff()).build();
    this.retry = Retry.of("GCRRegistry", config);
  }`

**Review by vtxorxwitty — APPROVED** [review]
> Reviewed 1 of 1 files at r2, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41804)* status: 2 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by sribalij — APPROVED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41804)* status: 1. unresolved discussions.

**Review by shivamnegi94 — COMMENTED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41804)* status: 2. unresolved discussions. 

___
*[`960-api-services/src/main/java/io/harness/artifacts/gcr/service/GcrApiServiceImpl.java` line 251 at r4](https://reviewable.io/reviews/harness/harness-core/41804#-NJiu0O_a0zb3FpUQzAG-r4-251:-NL-A8zw8cRJ5Z6o7tWX:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/160a50356b4a5fd4d6412aa1f8003e94b225f401/960-api-services/src/main/java/io/harness/artifacts/gcr/service/GcrApiServiceImpl.java#L251)):*
<details><summary><i>Previously, sribalij (Sridhar) wrote…</i></summary><blockquote>

This method `getBuilds(GcrRestClient gcrRestClient, String basicAuthHeader, String imageName)` and
`listImageTag(GcrRestClient gcrRestClient, String basicAuthHeader, String imageName) throws Exception {`
are doing the same thing right except for return type? Can the same thing be used?
</blockquote></details>

Done.

**Review by sribalij — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41804)* status: :shipit: complete! ✓

**Pull request merged**